### PR TITLE
Limit fusion to ops with single uses to avoid duplication

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

**Comment by benvanik** [comment]
(curious about a benchmark run here - this may create a lot more dispatch regions as something that could have fused becomes its own)

**Comment by ThomasRaoux** [comment]
> (curious about a benchmark run here - this may create a lot more dispatch regions as something that could have fused becomes its own)

Yes I'm running the benchmark to see the result. On Bert training it prevents the gather which has high cost to be duplicated. It creates quite a few more dispatch but overall helps perf. I'll refine it if needed based on benchmark results.

File: iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp [hunk 1]
```diff
@@ -79,12 +79,14 @@ struct FusionOfTensorOpsPass
...
-          for (Operation *user : producer.getUsers()) {
-            if (isa<linalg::GenericOp>(user)) continue;
-            numUsers.insert(user);
-          }
-          return numUsers.empty();
+          // Only fuse if it has a single linalg generic user. It is a
+          // simplistic heuristic to avoid duplicating ops that may be
+          // expensive.
+          // TODO: Add a cost model to allow ops to be duplicated.
+          if (std::distance(producer.getUsers().begin(),
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:86]: I think you can use `llvm::hasSingleElement(producer.getUsers())`

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp]: done.

**Pull request closed without merging**